Source/EphysSocket.cpp

```diff
@@ -162,51 +164,106 @@ bool EphysSocket::startAcquisition()
     total_samples = 0;
     eventState = 0;
 
+    error_flag = false;
+    stop_flag = false;
+    full_flag = false;
+    buffer_flag = false;
+
+    Thread::launch([this] {runBufferThread(); });
     startThread();
 
     return true;
 }
 
-
 bool EphysSocket::stopAcquisition()
 {
     if (isThreadRunning())
     {
         signalThreadShouldExit();
     }
 
+    stop_flag = true;
+
     waitForThreadToExit(500);
 
     sourceBuffers[0]->clear();
     return true;
 }
 
-bool EphysSocket::updateBuffer()
+void EphysSocket::runBufferThread()
 {
-    int rc = socket->read(recvbuf, num_channels * num_samp * 2, true);
+    const int header_size = sizeof(int) * 2;
+    const int total_samples = num_channels * num_samp;
+    const int total_packet_size = total_samples * sizeof(uint16_t);
+    const int packet_ratio = ((total_packet_size + header_size) / MAX_PACKET_SIZE) + 1;
+    const int packet_size = (total_packet_size / packet_ratio) + header_size;
+
+    std::vector<uint16_t> read_buffer;
+    read_buffer.resize(total_samples + 4);
 
-    if (rc == -1)
+    int rc = 0, offset = 0, bytes_sent = 0;
+
+    while (!stop_flag)
     {
-        CoreServices::sendStatusMessage("Ephys Socket: Data shape mismatch");
-        return false;
-    }
-   
-    int k = 0;
-    for (int i = 0; i < num_samp; i++) {
-        for (int j = 0; j < num_channels; j++) {
-            convbuf[k++] = data_scale *  (float)(recvbuf[j*num_samp + i] - data_offset);
+        rc = socket->read(read_buffer.data(), packet_size, true);
+
+        if (rc == -1)
+        {
+            CoreServices::sendStatusMessage("Ephys Socket: Data shape mismatch");
+            error_flag = true;
+            return;
+        }
+
+        offset = read_buffer.at(1) << 16 | read_buffer.at(0);
+        bytes_sent = read_buffer.at(3) << 16 | read_buffer.at(2);
+
+        if (!buffer_flag) recvbuf0.insert(recvbuf0.begin() + (offset / sizeof(uint16_t)), read_buffer.begin() + 4, read_buffer.end()); // This might error if packets are lost
```

Comment:
If statements without brackets are ok for things that would reasonably fit as a ternary operator but when there is no assignment. These statements are too long for that. Please use squigglies.

Reply:
`std::vector::insert` is for prepending an array dynamically. I'm not sure that's what you want. I think you have pre-allocated these arrays so what you really want is `std::vector:;assign`

Reply:
This also caused a hard crash when i set the buffer size parameters incorrectly. Make sure you add nonsense values into the GUI to ensure that exceptions are properly handled.

Reply:
Given all this might be worth putting the size of the matrix into the header. We are implicitly assuming some size information by sending the index, so its no less clean.

Reply:
Currently I don't pre-allocate the data itself, only the size of the container. I am using `std::vector::insert` because there might be multiple packets, and using `std::vector::assign` would remove any previously added data elements. I couldn't find a method in `std::vector` that allows for overwriting particular blocks of the container, but I may have missed it.

Reply:
`std::copy` overwrites existing data in a vector, this is implemented now.